---
layout: default
title:  Reading materials
---

#### <img border="0" src="https://www.svgrepo.com/show/26916/book.svg" width="25" height="25"> Reading materials  
This page contains recommended reading materials in addition to the [pre-course content][1]. 

**!** - important

**Machine Learning approaches in Integration**  
Note: if you cannot see one of the pages because you've reached the free member preview,
open the page in "Private Browsing" mode in your web browser.
- **!** [Feature selection][17] 
- **!** [Dimensionality reduction][18] 
- **!** [Supervised Omics integration][19] 
- **!** [Unsupervised Omics integration][20] 
- [Multi-omics approaches to disease][21]
  <br/><br/>

**Biological Network Analysis and Network Topology**
- **!** [Network biology: understanding the cell's functional organization][13] 
- **!** [Network medicine: a network-based approach to human disease][14] 
- [Network biology concepts in complex disease comorbidities][15]
- [Gene co-expression analysis and conserved modules][16]
  <br/><br/>

**Genome-scale metabolic modeling**
- **!** [Overview of genome-scale metabolic models][2]
- [What is flux balance analysis?][3]
- [A Systematic Evaluation of Methods for Tailoring Genome-Scale Metabolic Models][4]
- Human Metabolic Models: [Human1][5] and [Recon3D][6]
- **!** [Reporter metabolite analysis][7]
- [Gene set enrichment analysis (GSEA)][8] and [Directional gene set analysis (GSA)][9]
- Modeling software: Python: [COBRApy][10], MATLAB: [RAVEN][11] and [COBRA][12]
  <br/><br/>

[1]: precourse.md
[2]: https://www.frontiersin.org/articles/10.3389/fphys.2015.00413/full
[3]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3108565/
[4]: https://www.cell.com/fulltext/S2405-4712(17)30010-8
[5]: https://pubmed.ncbi.nlm.nih.gov/32209698/
[6]: https://www.ncbi.nlm.nih.gov/pubmed/29457794
[7]: https://www.pnas.org/content/102/8/2685
[8]: https://www.pnas.org/content/102/43/15545
[9]: https://pubmed.ncbi.nlm.nih.gov/23444143/
[10]: https://opencobra.github.io/cobrapy/
[11]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6207324/
[12]: https://www.nature.com/articles/s41596-018-0098-2
[13]: https://www.nature.com/articles/nrg1272
[14]: https://www.nature.com/articles/nrg2918
[15]: https://www.nature.com/articles/nrg.2016.87.pdf
[16]: https://science.sciencemag.org/content/302/5643/249.full
[17]: https://towardsdatascience.com/select-features-for-omics-integration-511390b7e7fd
[18]: https://towardsdatascience.com/reduce-dimensions-for-single-cell-4224778a2d67
[19]: https://towardsdatascience.com/supervised-omics-integration-2158e1a6d23f
[20]: https://medium.com/@nikolay.oskolkov/unsupervised-omics-integration-688bf8fa49bf
[21]: https://link.springer.com/content/pdf/10.1186/s13059-017-1215-1.pdf

[Back to Homepage](https://nbisweden.github.io/workshop_omics_integration/)
