{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Proteomics Integration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load a enzyme constrained metabolic mdoel of _Escherichia coli_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from cobra.io import read_sbml_model\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "\n",
    "model = read_sbml_model('data/eciML1515.xml.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The model has two differences with a standard COBRA model. First, the reactions contain another _metabolite_: the enyzme itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>FRD2No1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Fumarate reductase (No1)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fedae3534e0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>fum_c + mql8_c + 1.1111e-06 prot_P00363 + 1.1111e-06 prot_P0A8Q0 + 1.1111e-06 prot_P0A8Q3 + 1.1111e-06 prot_P0AC47 --> mqn8_c + succ_c</p>\n",
       "                    <p style='text-align:right'>Fumarate [cytosol] + Menaquinol 8 [cytosol] + 1.1111e-06 prot_P00363 [cytosol] + 1.1111e-06 prot_P0A8Q0 [cytosol] + 1.1111e-06 prot_P0A8Q3 [cytosol] + 1.1111e-06 prot_P0AC47 [cytosol] --> Menaquino...</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>b4153 and b4152 and b4151 and b4154</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction FRD2No1 at 0x7fedae3534e0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.FRD2No1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this model, all protein ids follow the form `prot_UNIPROT`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second difference is the existence of _protein exchange reactions_. These protein exchanges follow the naming `prot_UNIPROT_exchange`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>prot_P00363_exchange</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>prot_P00363_exchange</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fedacdccfd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'> --> prot_P00363</p>\n",
       "                    <p style='text-align:right'> --> prot_P00363 [cytosol]</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>b4154</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction prot_P00363_exchange at 0x7fedacdccfd0>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.prot_P00363_exchange"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By putting an upper bound on these exchanges, we will integrate proteomics data into the model and treat it as an usual COBRA model without further changes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Proteomics data preparation\n",
    "Proteomics data is usually presented as Number of copies per cell. We need to do convert this information into the units used by the enzyme constrained model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>DEBUG:</b> Should run PCA for the entire dataset?\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    #\"data/ecoli_proteomics_schmidt2016_S9.tsv\", # our strain but just Glc/LB\n",
    "    \"data/ecoli_proteomics_schmidt2016_S5_ren.tsv\", # BW25113 22 media\n",
    "    \"\\t\", skiprows=2  # skip titles and subtitles (XLXS)\n",
    ")\n",
    "exp_details = pd.read_csv(\n",
    "    \"data/ecoli_details_schmidt2016_S23.tsv\", \"\\t\", \n",
    "    skiprows=2  # skip titles and subtitles (XLXS)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Peptides used for quantitation</th>\n",
       "      <th>Confidence score</th>\n",
       "      <th>Molecular weight (Da)</th>\n",
       "      <th>Glucose_copies</th>\n",
       "      <th>LB_copies</th>\n",
       "      <th>Glycerol + AA_copies</th>\n",
       "      <th>Acetate_copies</th>\n",
       "      <th>...</th>\n",
       "      <th>Stationary phase 1 day_cv</th>\n",
       "      <th>Stationary phase 3 days_cv</th>\n",
       "      <th>Osmotic-stress glucose_cv</th>\n",
       "      <th>42°C glucose_cv</th>\n",
       "      <th>pH6 glucose_cv</th>\n",
       "      <th>Xylose_cv</th>\n",
       "      <th>Mannose_cv</th>\n",
       "      <th>Galactose _cv</th>\n",
       "      <th>Succinate_cv</th>\n",
       "      <th>Fructose_cv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P0A8T7</td>\n",
       "      <td>DNA-directed RNA polymerase subunit beta' OS=E...</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>91</td>\n",
       "      <td>6045.53</td>\n",
       "      <td>155045.0080</td>\n",
       "      <td>2779</td>\n",
       "      <td>7164</td>\n",
       "      <td>4503</td>\n",
       "      <td>2180</td>\n",
       "      <td>...</td>\n",
       "      <td>6.76</td>\n",
       "      <td>12.08</td>\n",
       "      <td>20.91</td>\n",
       "      <td>13.86</td>\n",
       "      <td>6.14</td>\n",
       "      <td>6.27</td>\n",
       "      <td>16.15</td>\n",
       "      <td>22.87</td>\n",
       "      <td>16.45</td>\n",
       "      <td>9.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P0A8V2</td>\n",
       "      <td>DNA-directed RNA polymerase subunit beta OS=Es...</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>89</td>\n",
       "      <td>5061.29</td>\n",
       "      <td>150520.2758</td>\n",
       "      <td>3957</td>\n",
       "      <td>8888</td>\n",
       "      <td>5199</td>\n",
       "      <td>2661</td>\n",
       "      <td>...</td>\n",
       "      <td>14.42</td>\n",
       "      <td>11.18</td>\n",
       "      <td>17.48</td>\n",
       "      <td>10.51</td>\n",
       "      <td>5.93</td>\n",
       "      <td>4.27</td>\n",
       "      <td>13.51</td>\n",
       "      <td>19.75</td>\n",
       "      <td>13.60</td>\n",
       "      <td>7.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P36683</td>\n",
       "      <td>Aconitate hydratase 2 OS=Escherichia coli (str...</td>\n",
       "      <td>acnB</td>\n",
       "      <td>67</td>\n",
       "      <td>4505.67</td>\n",
       "      <td>93420.9457</td>\n",
       "      <td>7596</td>\n",
       "      <td>16600</td>\n",
       "      <td>17548</td>\n",
       "      <td>22844</td>\n",
       "      <td>...</td>\n",
       "      <td>3.93</td>\n",
       "      <td>10.04</td>\n",
       "      <td>16.60</td>\n",
       "      <td>9.57</td>\n",
       "      <td>4.47</td>\n",
       "      <td>21.78</td>\n",
       "      <td>3.16</td>\n",
       "      <td>2.07</td>\n",
       "      <td>4.21</td>\n",
       "      <td>3.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P15254</td>\n",
       "      <td>Phosphoribosylformylglycinamidine synthase OS=...</td>\n",
       "      <td>purL</td>\n",
       "      <td>65</td>\n",
       "      <td>4277.71</td>\n",
       "      <td>141295.8984</td>\n",
       "      <td>2456</td>\n",
       "      <td>821</td>\n",
       "      <td>2339</td>\n",
       "      <td>1438</td>\n",
       "      <td>...</td>\n",
       "      <td>7.01</td>\n",
       "      <td>4.02</td>\n",
       "      <td>12.97</td>\n",
       "      <td>11.38</td>\n",
       "      <td>1.78</td>\n",
       "      <td>7.88</td>\n",
       "      <td>1.22</td>\n",
       "      <td>7.69</td>\n",
       "      <td>2.96</td>\n",
       "      <td>6.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P09831</td>\n",
       "      <td>Glutamate synthase [NADPH] large chain OS=Esch...</td>\n",
       "      <td>gltB</td>\n",
       "      <td>64</td>\n",
       "      <td>4111.74</td>\n",
       "      <td>163176.3153</td>\n",
       "      <td>2859</td>\n",
       "      <td>604</td>\n",
       "      <td>652</td>\n",
       "      <td>1363</td>\n",
       "      <td>...</td>\n",
       "      <td>17.44</td>\n",
       "      <td>3.98</td>\n",
       "      <td>28.41</td>\n",
       "      <td>17.39</td>\n",
       "      <td>7.84</td>\n",
       "      <td>20.05</td>\n",
       "      <td>17.71</td>\n",
       "      <td>22.36</td>\n",
       "      <td>10.69</td>\n",
       "      <td>8.91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Uniprot Accession                                        Description  Gene  \\\n",
       "0            P0A8T7  DNA-directed RNA polymerase subunit beta' OS=E...  rpoC   \n",
       "1            P0A8V2  DNA-directed RNA polymerase subunit beta OS=Es...  rpoB   \n",
       "2            P36683  Aconitate hydratase 2 OS=Escherichia coli (str...  acnB   \n",
       "3            P15254  Phosphoribosylformylglycinamidine synthase OS=...  purL   \n",
       "4            P09831  Glutamate synthase [NADPH] large chain OS=Esch...  gltB   \n",
       "\n",
       "   Peptides used for quantitation  Confidence score  Molecular weight (Da)  \\\n",
       "0                              91           6045.53            155045.0080   \n",
       "1                              89           5061.29            150520.2758   \n",
       "2                              67           4505.67             93420.9457   \n",
       "3                              65           4277.71            141295.8984   \n",
       "4                              64           4111.74            163176.3153   \n",
       "\n",
       "   Glucose_copies  LB_copies  Glycerol + AA_copies  Acetate_copies  ...  \\\n",
       "0            2779       7164                  4503            2180  ...   \n",
       "1            3957       8888                  5199            2661  ...   \n",
       "2            7596      16600                 17548           22844  ...   \n",
       "3            2456        821                  2339            1438  ...   \n",
       "4            2859        604                   652            1363  ...   \n",
       "\n",
       "   Stationary phase 1 day_cv  Stationary phase 3 days_cv  \\\n",
       "0                       6.76                       12.08   \n",
       "1                      14.42                       11.18   \n",
       "2                       3.93                       10.04   \n",
       "3                       7.01                        4.02   \n",
       "4                      17.44                        3.98   \n",
       "\n",
       "   Osmotic-stress glucose_cv  42°C glucose_cv  pH6 glucose_cv  Xylose_cv  \\\n",
       "0                      20.91            13.86            6.14       6.27   \n",
       "1                      17.48            10.51            5.93       4.27   \n",
       "2                      16.60             9.57            4.47      21.78   \n",
       "3                      12.97            11.38            1.78       7.88   \n",
       "4                      28.41            17.39            7.84      20.05   \n",
       "\n",
       "   Mannose_cv  Galactose _cv  Succinate_cv  Fructose_cv  \n",
       "0       16.15          22.87         16.45         9.29  \n",
       "1       13.51          19.75         13.60         7.77  \n",
       "2        3.16           2.07          4.21         3.25  \n",
       "3        1.22           7.69          2.96         6.20  \n",
       "4       17.71          22.36         10.69         8.91  \n",
       "\n",
       "[5 rows x 72 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Growth condition</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Growth rate (h-1)</th>\n",
       "      <th>Stdev</th>\n",
       "      <th>Single cell volume [fl]1</th>\n",
       "      <th>Doubling time (h-1)</th>\n",
       "      <th>Time exp before harvest (h)</th>\n",
       "      <th># of doublings at exponential growth before harvesting</th>\n",
       "      <th>OD @ harvesting. replicates</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Number of Proteins Identified (FDR 1%)2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LB</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>1.90</td>\n",
       "      <td>0.03</td>\n",
       "      <td>4.29</td>\n",
       "      <td>0.4</td>\n",
       "      <td>5</td>\n",
       "      <td>13.7</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1.37</td>\n",
       "      <td>1.34</td>\n",
       "      <td>1752.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LB</td>\n",
       "      <td>MG1665</td>\n",
       "      <td>1.78</td>\n",
       "      <td>0.05</td>\n",
       "      <td>4.23</td>\n",
       "      <td>0.4</td>\n",
       "      <td>5.4</td>\n",
       "      <td>13.9</td>\n",
       "      <td>0.53</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.32</td>\n",
       "      <td>1733.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LB</td>\n",
       "      <td>NCM3722</td>\n",
       "      <td>2.30</td>\n",
       "      <td>0.05</td>\n",
       "      <td>4.36</td>\n",
       "      <td>0.3</td>\n",
       "      <td>3.2</td>\n",
       "      <td>10.6</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1703.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glycerol + AA</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>1.27</td>\n",
       "      <td>0.01</td>\n",
       "      <td>3.83</td>\n",
       "      <td>0.5</td>\n",
       "      <td>7.5</td>\n",
       "      <td>13.8</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1675.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2.30</td>\n",
       "      <td>2.3</td>\n",
       "      <td>29.3</td>\n",
       "      <td>12.7</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1683.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Growth condition   Strain  Growth rate (h-1)  Stdev  \\\n",
       "0               LB  BW25113               1.90   0.03   \n",
       "1               LB   MG1665               1.78   0.05   \n",
       "2               LB  NCM3722               2.30   0.05   \n",
       "3    Glycerol + AA  BW25113               1.27   0.01   \n",
       "4          Acetate  BW25113               0.30   0.04   \n",
       "\n",
       "   Single cell volume [fl]1 Doubling time (h-1) Time exp before harvest (h)  \\\n",
       "0                      4.29                 0.4                           5   \n",
       "1                      4.23                 0.4                         5.4   \n",
       "2                      4.36                 0.3                         3.2   \n",
       "3                      3.83                 0.5                         7.5   \n",
       "4                      2.30                 2.3                        29.3   \n",
       "\n",
       "  # of doublings at exponential growth before harvesting  \\\n",
       "0                                               13.7       \n",
       "1                                               13.9       \n",
       "2                                               10.6       \n",
       "3                                               13.8       \n",
       "4                                               12.7       \n",
       "\n",
       "   OD @ harvesting. replicates  Unnamed: 9  Unnamed: 10  \\\n",
       "0                         1.80        1.37         1.34   \n",
       "1                         0.53        1.09         1.32   \n",
       "2                         0.91        0.96         1.00   \n",
       "3                         0.50        0.51         0.50   \n",
       "4                         0.53        0.22         0.48   \n",
       "\n",
       "   Number of Proteins Identified (FDR 1%)2  \n",
       "0                                   1752.0  \n",
       "1                                   1733.0  \n",
       "2                                   1703.0  \n",
       "3                                   1675.0  \n",
       "4                                   1683.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_details.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[:, \n",
    "    df.columns.str.contains(\"_copies|_cv\", regex=True) |  # only interested in copies/cell and uncertainty\n",
    "    df.columns.isin([\"Uniprot Accession\"])  # and relevant info about proteins\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_ac = df.loc[:, [\"Uniprot Accession\", \"Acetate_copies\", \"Acetate_cv\"]]\n",
    "# rename resulting columns\n",
    "df_ac.columns = [\"uniprot\", \"copies_per_cell\", \"CV\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>copies_per_cell</th>\n",
       "      <th>CV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2058.000000</td>\n",
       "      <td>2058.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1847.405248</td>\n",
       "      <td>17.656040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8487.943632</td>\n",
       "      <td>20.773887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.190000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>29.000000</td>\n",
       "      <td>6.680000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>170.000000</td>\n",
       "      <td>11.425000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>868.250000</td>\n",
       "      <td>19.892500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>242737.000000</td>\n",
       "      <td>170.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       copies_per_cell           CV\n",
       "count      2058.000000  2058.000000\n",
       "mean       1847.405248    17.656040\n",
       "std        8487.943632    20.773887\n",
       "min           0.000000     0.190000\n",
       "25%          29.000000     6.680000\n",
       "50%         170.000000    11.425000\n",
       "75%         868.250000    19.892500\n",
       "max      242737.000000   170.100000"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ac.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply uncertainty (extend upper bound as 1/2 of stdev)\n",
    "df_ac[\"copies_upper\"] = df_ac[\"copies_per_cell\"] + 0.5 * df_ac[\"CV\"]/100 * df_ac[\"copies_per_cell\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, convert copies per cell to abundance per cell.\n",
    "\\begin{align}\n",
    "\\frac{\\text{mmol}}{\\text{cell}} = \\frac{\\text{molecules}}{\\text{cell}} \\frac{10^3\\text{mol}}{\\text{molecules}}\n",
    "\\end{align}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ac[\"mmol_per_cell\"] = df_ac[\"copies_upper\"] * 1e3/6.022e23"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And, then, convert the abundance per cell into abundance per gDW.\n",
    "\n",
    "\\begin{align}\n",
    "\\frac{\\text{mmol}}{\\text{gDW}} = \\frac{\\text{mmol}}{\\text{cell}} \\frac{\\text{cell}}{fL} \\frac{fL}{g}\\frac{g}{\\text{gDW}}\n",
    "\\end{align}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "growth_experimental = exp_details.loc[\n",
    "    (exp_details[\"Growth condition\"]==\"Acetate\") & (exp_details[\"Strain\"]==\"BW25113\"), \n",
    "    \"Growth rate (h-1)\"\n",
    "].values[0]\n",
    "cell_volume = exp_details.loc[\n",
    "    (exp_details[\"Growth condition\"]==\"Acetate\") & (exp_details[\"Strain\"]==\"BW25113\"), \n",
    "    \"Single cell volume [fl]1\"\n",
    "].values[0]\n",
    "cell_density = 1.105e-12\n",
    "water_content = 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ac[\"conc\"] = df_ac[\"mmol_per_cell\"] * 1 / (cell_volume * cell_density * water_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics = df_ac[\"conc\"]\n",
    "proteomics.index = df_ac[\"uniprot\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Model\n",
    "Simulations part (caffeine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the non enzyme-constrained model\n",
    "plain_model = model.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def limit_proteins(model, measurements):\n",
    "    \"\"\"Apply proteomics measurements to `model`.\n",
    "\n",
    "    Adapted from https://github.com/DD-DeCaF/simulations/blob/devel/src/simulations/modeling/driven.py\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: cobra.Model\n",
    "        The enzyme-constrained model.\n",
    "    measurements : pd.DataFrame\n",
    "        Protein abundances in mmol / gDW.\n",
    "\n",
    "    \"\"\"\n",
    "    for protein_id, measure in measurements.items():\n",
    "        try:\n",
    "            rxn = model.reactions.get_by_id(f\"prot_{protein_id}_exchange\")\n",
    "        except KeyError:\n",
    "            pass\n",
    "        else:\n",
    "            # update only upper_bound (as enzymes can be unsaturated):\n",
    "            rxn.bounds = (0, measure)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Optimize the enzyme constrained model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "limit_proteins(model, proteomics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.000</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_acgam_e</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cellb_e</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_chol_e</th>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_pi_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h_e</th>\n",
       "      <td>6.682729</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q59385_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BEX0_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF16_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF17_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q93K97_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6083 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.000 at 0x7fedab946a58>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# enzyme-contrained (the model doesn't grow)\n",
    "\n",
    "model.optimize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The model can't grow!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Flexibilization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Experimental measurements can be too restrictive if an uncertainty is not given. Thus, a flexibilization of the proteomics data is usually required to work with enzyme constrained models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_shadow_prices(solution, met_ids, top=1):\n",
    "    \"\"\"\n",
    "    Retrieves shadow prices for a list of metabolites from the solution and ranks\n",
    "    them from most to least sensitive in the model.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    solution: cobra.Solution\n",
    "        The usual Solution object returned by model.optimize().\n",
    "    biomass_reaction: str\n",
    "        name of biomass reaction\n",
    "    met_ids: iterable of strings\n",
    "        Subset of metabolite IDs from the model.\n",
    "    top: int\n",
    "        The number of metabolites to be returned.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    shadow_pr: pd.Series\n",
    "        Top shadow prices, ranked.\n",
    "    \"\"\"\n",
    "    shadow_pr = solution.shadow_prices\n",
    "    shadow_pr = shadow_pr.loc[shadow_pr.index.isin(met_ids)]\n",
    "    return shadow_pr.sort_values()[:top]\n",
    "\n",
    "\n",
    "def flexibilize_proteomics(\n",
    "    model, biomass_reaction, minimal_growth, proteomics\n",
    "):\n",
    "    \"\"\"\n",
    "    Replace proteomics measurements with a set that enables the model to grow. Proteins\n",
    "    are removed from the set iteratively based on sensitivity analysis (shadow prices).\n",
    "    \n",
    "    Adapted from https://github.com/DD-DeCaF/simulations/blob/devel/src/simulations/modeling/driven.py\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: cobra.Model\n",
    "        The enzyme-constrained model.\n",
    "    minimal_growth_rate: float\n",
    "        Minimal growth rate to enforce.\n",
    "    proteomics: pandas.DataFrame\n",
    "        List of measurements.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    growth_rate: dict\n",
    "        New growth rate (will change if the model couldn't grow at the inputted value).\n",
    "    proteomics: list(dict)\n",
    "        Filtered list of proteomics.\n",
    "\n",
    "    \"\"\"\n",
    "    def protein_to_metabolite(protein_id, model):\n",
    "        met_id = model.metabolites.query(lambda m: protein_id in m.id)\n",
    "        return met_id[0].id if met_id else \"\"\n",
    "    \n",
    "    # reset growth rate in model:\n",
    "    model.reactions.get_by_id(biomass_reaction).bounds = (0, 1000)\n",
    "\n",
    "    # build a table with protein ids, met ids in model and values to constrain with:\n",
    "    prot_df = pd.DataFrame(proteomics)\n",
    "    prot_df.index = prot_df.index.astype(\"str\")\n",
    "    prot_df[\"met_id\"] = [protein_to_metabolite(prot, model) for prot in prot_df.index]\n",
    "    prot_df = prot_df[prot_df.met_id != \"\"]\n",
    "    \n",
    "    # constrain the model with all proteins and optimize:\n",
    "    limit_proteins(model, proteomics)\n",
    "    solution = model.optimize()\n",
    "    new_growth_rate = solution.objective_value if solution.objective_value else 0\n",
    "    \n",
    "    # relax growth constraint\n",
    "    minimal_growth *= 1.05\n",
    "\n",
    "    # while the model cannot grow to the desired level, remove the protein with\n",
    "    # the highest shadow price:\n",
    "    prots_to_remove = []\n",
    "    while new_growth_rate < minimal_growth and not prot_df.empty:\n",
    "        # get most influential protein in model:\n",
    "        top_protein = top_shadow_prices(solution, list(prot_df[\"met_id\"]))\n",
    "        top_protein = top_protein.index[0]\n",
    "        top_protein = prot_df.index[prot_df[\"met_id\"] == top_protein][0]\n",
    "\n",
    "        # update data: append protein to list, remove from current dataframe and\n",
    "        # increase the corresponding upper bound to +1000:\n",
    "        prots_to_remove.append(top_protein)\n",
    "        prot_df = prot_df.drop(labels=top_protein)\n",
    "        limit_proteins(model, pd.Series(data=[1000], index=[top_protein]))\n",
    "\n",
    "        # re-compute solution:\n",
    "        solution = model.optimize()\n",
    "        #if solution.objective_value == new_growth_rate:  # the algorithm is stuck\n",
    "        #    break\n",
    "        new_growth_rate = solution.objective_value if solution.objective_value else 0\n",
    "\n",
    "    # update growth rate if optimization was not successful:\n",
    "    if new_growth_rate < minimal_growth:\n",
    "        print(\n",
    "            f\"Minimal growth was not reached! \"\n",
    "            f\"Final growth of the model: {new_growth_rate}\"\n",
    "        )\n",
    "\n",
    "    return new_growth_rate, prots_to_remove"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enforce 0.1 of growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "biomass_reaction = \"BIOMASS_Ec_iML1515_core_75p37M\"\n",
    "new_growth_rate, prots_removed = flexibilize_proteomics(model, biomass_reaction, growth_experimental, proteomics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteins in dataset: 2058\n",
      "Proteins removed: 90\n"
     ]
    }
   ],
   "source": [
    "print(f\"Proteins in dataset: {proteomics.shape[0]}\\nProteins removed: {len(prots_removed)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.318</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_acgam_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.162421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cellb_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.121816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_chol_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_pi_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h_e</th>\n",
       "      <td>17.879146</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q59385_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BEX0_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF16_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF17_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q93K97_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6083 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.318 at 0x7fedab904c18>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.877</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_acgam_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.238866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cellb_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.361121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_chol_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.022570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_pi_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h_e</th>\n",
       "      <td>8.058201</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q59385_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BEX0_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF16_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q6BF17_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_Q93K97_exchange</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6083 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.877 at 0x7fedab904ac8>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plain_model.optimize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare the carbon source utilization of both models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exchanges in this are on the right-had side\n",
    "plain_exchanges = [reaction for reaction in plain_model.exchanges if reaction.flux > 0]\n",
    "enzyme_exchanges = [reaction for reaction in model.exchanges if reaction.flux > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "plain_fva = flux_variability_analysis(plain_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n"
     ]
    }
   ],
   "source": [
    "enzyme_fva = flux_variability_analysis(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MDH_REVNo1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.402620e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDHNo1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.339076e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK_REVNo1</th>\n",
       "      <td>18.914082</td>\n",
       "      <td>1.221030e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PUNP5No2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.105349e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGKNo1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.029215e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUAt2pp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.391933e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLYt2pp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.391991e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FORt2pp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.395550e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALAt4pp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.403159e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPAt4pp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-3.418356e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5069 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              minimum       maximum\n",
       "MDH_REVNo1   0.000000  2.402620e+02\n",
       "MDHNo1       0.000000  2.339076e+02\n",
       "PGK_REVNo1  18.914082  1.221030e+02\n",
       "PUNP5No2     0.000000  1.105349e+02\n",
       "PGKNo1       0.000000  1.029215e+02\n",
       "...               ...           ...\n",
       "GUAt2pp      0.000000 -2.391933e-08\n",
       "GLYt2pp      0.000000 -2.391991e-08\n",
       "FORt2pp      0.000000 -2.395550e-08\n",
       "ALAt4pp      0.000000 -2.403159e-08\n",
       "PPAt4pp      0.000000 -3.418356e-08\n",
       "\n",
       "[5069 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzyme_fva[enzyme_fva.maximum < 700].sort_values(\"maximum\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>prot_P21177_exchange</th>\n",
       "      <td>5.835889e-07</td>\n",
       "      <td>5.658190e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0AB71_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.695348e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0A991_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.389096e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P33025_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.669400e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h2o_e</th>\n",
       "      <td>4.716230e+01</td>\n",
       "      <td>5.119140e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FDH4ppNo2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-3.859694e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HPYRRyNo2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.536509e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arm_HPYRRy</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.536509e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HPYRRyNo1</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.536509e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0AB65_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.617528e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4376 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           minimum       maximum\n",
       "prot_P21177_exchange  5.835889e-07  5.658190e+02\n",
       "prot_P0AB71_exchange  0.000000e+00  3.695348e+02\n",
       "prot_P0A991_exchange  0.000000e+00  1.389096e+02\n",
       "prot_P33025_exchange  0.000000e+00  6.669400e+01\n",
       "EX_h2o_e              4.716230e+01  5.119140e+01\n",
       "...                            ...           ...\n",
       "FDH4ppNo2             0.000000e+00 -3.859694e-10\n",
       "HPYRRyNo2             0.000000e+00 -4.536509e-10\n",
       "arm_HPYRRy            0.000000e+00 -4.536509e-10\n",
       "HPYRRyNo1             0.000000e+00 -4.536509e-10\n",
       "prot_P0AB65_exchange  0.000000e+00 -2.617528e-09\n",
       "\n",
       "[4376 rows x 2 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plain_fva[plain_fva.maximum < 700].sort_values(\"maximum\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>prot_P37769_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.314800e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0AC16_exchange</th>\n",
       "      <td>6.629624e-07</td>\n",
       "      <td>1.248440e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0AEK2_exchange</th>\n",
       "      <td>1.163420e-05</td>\n",
       "      <td>7.532970e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0AA84_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.924040e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P21151_exchange</th>\n",
       "      <td>1.244089e-04</td>\n",
       "      <td>2.031869e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P69330_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-1.542749e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0A9I1_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-1.543561e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P38135_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-1.958814e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0A746_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-3.368114e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_P0A8F4_exchange</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.164402e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1259 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           minimum       maximum\n",
       "prot_P37769_exchange  0.000000e+00  2.314800e+01\n",
       "prot_P0AC16_exchange  6.629624e-07  1.248440e+01\n",
       "prot_P0AEK2_exchange  1.163420e-05  7.532970e+00\n",
       "prot_P0AA84_exchange  0.000000e+00  2.924040e+00\n",
       "prot_P21151_exchange  1.244089e-04  2.031869e+00\n",
       "...                            ...           ...\n",
       "prot_P69330_exchange  0.000000e+00 -1.542749e-11\n",
       "prot_P0A9I1_exchange  0.000000e+00 -1.543561e-11\n",
       "prot_P38135_exchange  0.000000e+00 -1.958814e-11\n",
       "prot_P0A746_exchange  0.000000e+00 -3.368114e-10\n",
       "prot_P0A8F4_exchange  0.000000e+00 -4.164402e-10\n",
       "\n",
       "[1259 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzyme_fva[enzyme_fva.index.str.startswith(\"prot_\")].sort_values(\"maximum\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Identify enzymatic bottlenecks in the enzymed constrained model in **Acetate** as carbon source (shadow prices?).\n",
    "* Prepare and limit the model for the medium with Glucose\n",
    "* Does the model grow? If not, try flexibilizing the model.\n",
    "* Identify enzymatic bottlenecks in the enzymed constrained model in with **Glucose** as carbon source (shadow prices?).\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
